jobs: TestPauseReason failed

jobs.TestPauseReason [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/14449002?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/14449002?buildTab=artifacts#/) on master @ [91c1d5cbb5a56cd9c082d78814cd03c5471dddfa](https://github.com/cockroachdb/cockroach/commits/91c1d5cbb5a56cd9c082d78814cd03c5471dddfa):


```
=== RUN   TestPauseReason
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/e2483f27c4b073bcdd211db4d1f76316/logTestPauseReason2953628379
    test_log_scope.go:81: use -show-logs to present logs inline
    test_server_shim.go:144: cluster virtualization disabled in global scope due to issue: #76378 (expected label: C-bug)
    conditional_wrap.go:190: 
        pkg/jobs_test/pkg/jobs/jobs_test.go:3017: (TestPauseReason)
        	NOTICE: .JobRegistry() called via implicit interface ApplicationLayerInterface;
        HINT: consider using .ApplicationLayer().JobRegistry() instead.
    conditional_wrap.go:194: TIP: consider replacing the test server initialization from:
            ts, ... := serverutils.StartServer(t, ...)
            defer ts.Stopper().Stop(...)
        to:
            srv, ... := serverutils.StartServer(t, ...)
            defer srv.Stopper().Stop(...)
            ts := srv.ApplicationLayer()
        
        See also: https://go.crdb.dev/p/testserver-and-cluster-virtualization
    jobs_test.go:3105: condition failed to evaluate within 45s: from jobs_test.go:3064: still waiting for claim to clear
    panic.go:626: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/e2483f27c4b073bcdd211db4d1f76316/logTestPauseReason2953628379
--- FAIL: TestPauseReason (46.39s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #120046 jobs: TestPauseReason failed [C-test-failure O-robot P-3 T-jobs branch-release-23.2]
- #116538 jobs: TestPauseReason failed [A-disaster-recovery A-jobs C-test-failure O-robot P-2 T-jobs branch-release-23.2.0-rc]
</p>
</details>
/cc @cockroachdb/jobs @cockroachdb/disaster-recovery
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestPauseReason.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-36816

jobs.TestPauseReason [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/14884544?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/14884544?buildTab=artifacts#/) on master @ [1abcc00a451ba6a908b156d80cd6c34759ecbe34](https://github.com/cockroachdb/cockroach/commits/1abcc00a451ba6a908b156d80cd6c34759ecbe34):


```
=== RUN   TestPauseReason
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/e2483f27c4b073bcdd211db4d1f76316/logTestPauseReason1847003222
    test_log_scope.go:81: use -show-logs to present logs inline
    test_server_shim.go:144: cluster virtualization disabled in global scope due to issue: #76378 (expected label: C-bug)
    jobs_test.go:3117: condition failed to evaluate within 45s: from jobs_test.go:3076: still waiting for claim to clear
    panic.go:626: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/e2483f27c4b073bcdd211db4d1f76316/logTestPauseReason1847003222
--- FAIL: TestPauseReason (46.24s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #120046 jobs: TestPauseReason failed [C-test-failure O-robot P-3 T-jobs branch-release-23.2]
- #116538 jobs: TestPauseReason failed [A-disaster-recovery A-jobs C-test-failure O-robot P-2 T-jobs branch-release-23.2.0-rc]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestPauseReason.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

--- a/pkg/jobs/adopt.go
+++ b/pkg/jobs/adopt.go
@@ -481,6 +481,7 @@ func (r *Registry) maybeClearLease(job *Job, jobErr error) {
 	if jobErr == nil {
 		return
 	}
+	r.unregister(job.ID())
 	r.clearLeaseForJobID(job.ID(), r.db.Executor(), nil /* txn */)
 }
 

